src/lib/storage/lz4/lz4_encoder.hpp

```diff
@@ -122,48 +137,348 @@ class LZ4Encoder : public SegmentEncoder<LZ4Encoder> {
+      return pmr_vector<char>{};
+    }
+
+    DebugAssert(dictionary_size <= max_dictionary_size,
+                "Generated ZSTD dictionary in LZ4 compression is larger than "
+                "the memory allocated for it.");
+    dictionary.resize(dictionary_size);
+    dictionary.shrink_to_fit();
+
+    return dictionary;
+  }
+
+  /**
+   * When training a dictionary for strings, each row element is a sample. This is not done when training a
+   * dictionary for non-strings since zstd's dictionary is made for strings (and the samples can have different sizes).
+   * First, a dictionary is trained with the provided data and sample sizes. If this does not succeed, we try to
+   * increase the input size by repeating the values and adding larger sample sizes up to a certain limit. If this still
+   * fails or the input size is too small in general, a dictionary won't be trained and can't be used for compression.
+   * Compressing the blocks independently and without a dictionary hurts the compression ratio.
+   *
+   * @param values The input data that will be compressed (i.e., all strings concatenated).
+   * @param sample_sizes A vector of sample lengths. Each length corresponds to a substring in the values vector. These
+   *                     should correspond to the length of each row's value.
+   * @return The trained dictionary or in the case of failure an empty vector.
+   */
+  pmr_vector<char> _train_string_dictionary(const pmr_vector<char>& values, const pmr_vector<size_t>& sample_sizes) {
+    /**
+     * The recommended dictionary size is about 1/100th of size of all samples combined, but he size also has to be at
+     * least 1KB. Smaller dictionaries won't work.
+     */
+    auto max_dictionary_size = values.size() / 100;
+    max_dictionary_size = std::max(max_dictionary_size, _minimum_dictionary_size);
+
+    auto dictionary = pmr_vector<char>{values.get_allocator()};
+    size_t dictionary_size;
+
+    // If the input does not contain enough values, it won't be possible to train a dictionary for it.
+    if (values.size() < _minimum_value_size) {
+      return dictionary;
+    }
+
+    dictionary.resize(max_dictionary_size);
+    dictionary_size = ZDICT_trainFromBuffer(dictionary.data(), max_dictionary_size, values.data(), sample_sizes.data(),
+                                            static_cast<unsigned>(sample_sizes.size()));
+
+    // If the generation failed, try generating a dictionary with more input.
+    if (ZDICT_isError(dictionary_size)) {
+      return _train_string_dictionary_padded(values, sample_sizes, max_dictionary_size);
+    }
+
+    DebugAssert(dictionary_size <= max_dictionary_size,
+                "Generated ZSTD dictionary in LZ4 compression is larger than "
+                "the memory allocated for it.");
+
+    // Shrink the allocated dictionary size to the actual size.
+    dictionary.resize(dictionary_size);
+    dictionary.shrink_to_fit();
+
+    return dictionary;
+  }
```

Comment:
Could you reduce duplication with above by using `if constexpr` or alike?

Reply:
We removed the code duplication when we removed the padding of the input data.